fitbenchmarking/core/results_output.py

```diff
@@ -199,6 +201,287 @@ def preprocess_data(results: "list[FittingResult]"):
+                    results[result_ix].jacobian_tag = column_rename
+
+            elif tag_to_be_updated == 'hessian':
+                # Only hessian tag need to be renamed
+                unpacked_column_tag['hessian'] = \
+                    results[result_ix].hessian_tag = column_rename
+
+            # Update results tag
+            all_result_tags[ix]["col"] = ":".join([unpacked_column_tag[key]
+                                                   for key in col_order])
+
+    return results, all_result_tags
+
+
+def _find_tag_to_rename(all_result_tags,
+                        sm_summary,
+                        col_order,
+                        fallback_columns):
+    """
+    The function determines which of the jacobian, hessian or both
+    tags to rename for the fallback columns.
+
+    :param all_result_tags: A list of tags that can be used
+                            to sort the results
+    :type all_result_tag: list[dict[str, str]]
+    :param sm_summary: The col tags organized by software
+                       and minimizer tags
+    :type sm_summary: dict[list[str]]
+    :param col_order: The sort order of the col tags
+    :type col_order: list[str]
+    :param fallback_columns: The col tag of the fallback columns
+    :type fallback_columns: list[str]
+
+    :return: the tags to update organized by software
+             and minimizer
+    :rtype: dict[str]]
+    """
+    # Find rows
+    rows = {row['row'] for row in all_result_tags}
+
+    # Create dict to store tag update type
+    update_summary = {}
+
+    for sm, tags in sm_summary.items():
+
+        # Two types of tags for software and minimizer
+        # Update jacobian, hessian or both tags
+        if len(tags) == 2:
+            jac_tag = set()
+            hes_tag = set()
+            for tag in tags:
+                up = dict(zip(col_order, tag.split(":")))
+                jac_tag.add(up['jacobian'])
+                hes_tag.add(up['hessian'])
+            update_summary[sm] = 'jacobian' if len(jac_tag) != 1 \
+                and len(hes_tag) == 1 else \
+                ('hessian' if len(jac_tag) == 1
+                 and len(hes_tag) != 1 else 'both')
+
+        # Three types of tags for software and minimizer
+        # Update both tags
+        elif len(tags) == 3:
+            update_summary[sm] = 'both'
+
+        # Four types of tags for software and minimizer
+        # Update jacobian, hessian or both tags
+        elif len(tags) == 4:
+
+            for tag in tags:
+
+                # Unpack the tag
+                unpacked_column_tag = dict(zip(col_order, tag.split(":")))
+
+                # Find jacobian and hessian matches
+                matches_summary = {}
+                for check_tag in ['jacobian', 'hessian']:
+                    match_str = ':'. join(["[^:]*" if key == check_tag else
+                                           unpacked_column_tag[key]
+                                           for key in col_order])
```

Comment:
If both jac and hessian fallback, we will have `analytic:analytic` and `scipy:scipy`.

In this case, neither of the cols will have matches so we'll also need to check the version that looks like `"[^:]*:[^:]*"`

Reply:
```
        # Two types of tags for software and minimizer
        # Update jacobian, hessian or both tags
        if len(tags) == 2:
            jac_tag = set()
            hes_tag = set()
            for tag in tags:
                up = dict(zip(col_order, tag.split(":")))
                jac_tag.add(up['jacobian'])
                hes_tag.add(up['hessian'])
            update_summary[sm] = 'jacobian' if len(jac_tag) != 1 \
                and len(hes_tag) == 1 else \
                ('hessian' if len(jac_tag) == 1
                 and len(hes_tag) != 1 else 'both')
```

This part of the logic will be able to handle the case where both fallback. I added one of those scenarios to `edgecase6.json`.

- `{'row': 'p1', 'col': 's1:m4:j1:h1', 'cat': 'WeightedNLLSCostFunc', 'result_ix': 12}`
- `{'row': 'p2', 'col': 's1:m4:j1:h1', 'cat': 'WeightedNLLSCostFunc', 'result_ix': 13}`
- `{'row': 'p3', 'col': 's1:m4:j2:h2', 'cat': 'WeightedNLLSCostFunc', 'result_ix': 14}`
- `{'row': 'p4', 'col': 's1:m4:j2:h2', 'cat': 'WeightedNLLSCostFunc', 'result_ix': 15}`

In this case, the col_tags are renamed as  `'s1:m4:best_avaliable:best_avaliable'`.

I did not incorporate the suggested change last time because it was not being used subsequently to set the conditions. But let me see if I can clean up the logic further and simplify it now.

Reply:
I have simplified the logic of ` _find_tag_to_rename()` further.